controllers/imageclusterinstall_controller.go

```diff
@@ -537,41 +554,101 @@ func (r *ImageClusterInstallReconciler) SetupWithManager(mgr ctrl.Manager) error
 		Complete(r)
 }
 
-func (r *ImageClusterInstallReconciler) setBMHImage(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost, url string) error {
+func (r *ImageClusterInstallReconciler) getBmhDataImage(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost) *bmh_v1alpha1.DataImage {
+	if bmh == nil {
+		return nil
+	}
+	dataImage := bmh_v1alpha1.DataImage{}
+	key := client.ObjectKey{
+		Name:      bmh.Name,
+		Namespace: bmh.Namespace,
+	}
+	err := r.Get(ctx, key, &dataImage)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+
+	return &dataImage
+}
+
+func (r *ImageClusterInstallReconciler) isInspectionEnabled(bmh *bmh_v1alpha1.BareMetalHost) bool {
+	if bmh == nil {
+		return false
+	}
+
+	if bmh.ObjectMeta.Annotations != nil && bmh.ObjectMeta.Annotations[inspectAnnotation] != "disabled" {
+		return true
+	}
+
+	return false
+}
+
+func (r *ImageClusterInstallReconciler) createBMHDataImage(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost, ici *v1alpha1.ImageClusterInstall, url string) error {
+	dataImage := r.getBmhDataImage(ctx, bmh)
+
+	if dataImage == nil {
+		r.Log.Infof("creating new dataImage for BareMetalHost (%s/%s)", bmh.Name, bmh.Namespace)
+
+		// Name and namespace must match the ones in BMH
+		dataImage = &bmh_v1alpha1.DataImage{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      bmh.Name,
+				Namespace: bmh.Namespace,
+			},
+			Spec: bmh_v1alpha1.DataImageSpec{
+				URL: url,
+			},
+		}
+		err := controllerutil.SetControllerReference(bmh, dataImage, r.Client.Scheme())
+		if err != nil {
+			return fmt.Errorf("failed to set controller reference for dataImage due to %w", err)
+		}
+
+		err = r.Create(ctx, dataImage)
+		if err != nil {
+			return fmt.Errorf("failed to create dataImage due to %w", err)
+		}
+	}
+
 	patch := client.MergeFrom(bmh.DeepCopy())
 
 	dirty := false
 	if !bmh.Spec.Online {
 		bmh.Spec.Online = true
+		r.Log.Infof("enabling BareMetalHost (%s/%s) Online", bmh.Name, bmh.Namespace)
 		dirty = true
 	}
-	if bmh.Spec.Image == nil {
-		bmh.Spec.Image = &bmh_v1alpha1.Image{}
-		dirty = true
-	}
-	if bmh.Spec.Image.URL != url {
-		bmh.Spec.Image.URL = url
-		dirty = true
+
+	if bmh.Status.Provisioning.State == bmh_v1alpha1.StateAvailable {
+		if !bmh.Spec.ExternallyProvisioned {
+			r.Log.Infof("Setting BareMetalHost (%s/%s) ExternallyProvisioned spec", bmh.Name, bmh.Namespace)
+			bmh.Spec.ExternallyProvisioned = true
+			dirty = true
+		}
 	}
-	liveIso := "live-iso"
-	if bmh.Spec.Image.DiskFormat == nil || *bmh.Spec.Image.DiskFormat != liveIso {
-		bmh.Spec.Image.DiskFormat = &liveIso
+
+	if bmh.Status.Provisioning.State == bmh_v1alpha1.StateAvailable && bmh.Status.PoweredOn &&
+		setAnnotaitonIfNotExists(&bmh.ObjectMeta, rebootAnnotation, "true") {
+		r.Log.Infof("Adding reboot annotations to BareMetalHost (%s/%s)", bmh.Name, bmh.Namespace)
 		dirty = true
 	}
```

Comment:
I think what we need is this:
```suggestion
	if bmh.Status.PoweredOn{
		setAnnotaitonIfNotExists(&bmh.ObjectMeta, rebootAnnotation, "true") {
		r.Log.Infof("Adding reboot annotations to BareMetalHost (%s/%s)", bmh.Name, bmh.Namespace)
		dirty = true
	}
	if !bmh.Spec.online{
	        bmh.Spec.online = true
		r.Log.Infof("Setting BareMetalHost (%s/%s)  online to true", bmh.Name, bmh.Namespace)
		dirty = true
	}
```

Reply:
Why? I don't think you are right here, you should not reboot node all the time bmh.Status.PoweredOn is set to true. You will never start installation. 
We should reboot node only in very strictly designed flows, like here.
We reboot only in case node had inspection, passed validation and is powered on, so we want to reboot into disk.
Maybe we are missing the case were node was provided powered on and we want to reboot it with new data image  but we should not look only on PoweredOn flag for sure.
I believe this is optimization and can be done after , for sure better not to reboot when we should not

Reply:
So please add a comment that the code handles this case and make sure we set externally provisioned when when doing this.

Reply:
Seems that we should wrap the logic that reboots the BMH with a check of a condition that we set when we trigger the reboot (installation started...)

Reply:
I'm not sure about why to remove the state available condition?
We are handling the online flag few lines above this part regardless the power on status, so here it already should be enabled.

Reply:
Opened a ticket for next pr https://issues.redhat.com/browse/MGMT-18899, for this specific case